### ISSUE TITLE
MinusMover is not canonical?

### ISSUE THREAD

**Issue by dwhswenson** [choderalab/openpathsampling#280]
With the changes that made `MinusMover` based on filtering, it is no longer a `canonical` move. It should be.

Maybe the best approach is similar to what I implemented for `_is_ensemble_change_mover`? Normally use the computational approach, but allow a `PathMover` to have a `_is_canonical` attribute (normally set to `None`, which means "use the computational approach") which the user/developer can manually set to be `True` (or `False`) for special cases.

Related: why is `MinusMover` a subclass of `WrappedMover`? Shouldn't it be a subclass of `EnsembleFilterMover` (this removing one layer of `.mover` from changes)? Does it actually use any of the `WrappedMover` behaviors?

**Comment by jhprinz** [comment]
Correct. Since the used WrappedMover is just a single mover the canonical will traverse inside and return the SequentialMover as the canonical one. 
I think you are right, a flag to make it canonical is probably the best and also when reading the code the simplest to understand if you want to create a new mover.

The Wrapped is used to make the MinusMove Separately show up in the visualization. So Minus is just an empty wrapper to indicate the MinusMove and the minus consists of doing a Filter Step, etc... Otherwise the Minus is a filter and the filtering will not be mentioned. However, in terms of what it does both approached will be identical.

**Comment by dwhswenson** [comment]
I'll self-assign for adding the flag (it should be a relatively short PR). That should just involve having the flag property in each move (similar to the `is_ensemble_change_mover` property), and modifying `PathMoveChange.canonical` to have `if pmc.mover.is_canonical: return pmc` inside the while loop, right? 

On the inheritance, it sounds to me like the trade-off is as follows:

* *in favor of `MinusMover(WrappedMover)`*: visualization explicitly shows the filter stage
* *in favor of `MinusMover(FilterMover)`*: analysis is easier (`change.mover.mover.mover.movers[0]` is REALLY ugly -- although `change.mover.mover.movers[0]` is only slightly less ugly -- maybe a cleaner approach would be to give `MinusMover` a `movers` attribute which is the same as its `self.mover.mover.movers`?)

I'd say keep it as is for now, but keep this in mind as something we might someday want to revisit.

**Comment by jhprinz** [comment]
Good idea. Keep it that way for now.

Maybe we can have a shortcut for the MinusMove like `MinusMover.moves` for analysis? So we are independent of the choice of inheritance. The `.mover` would just be a property that returns `self.submover.submover.submovers` or so.

**Comment by dwhswenson** [comment]
>> maybe a cleaner approach would be to give MinusMover a movers attribute which is the same as its self.mover.mover.movers?

> Maybe we can have a shortcut for the MinusMove like MinusMover.moves for analysis?


I think we said the same thing. :-)

**Comment by jhprinz** [comment]
Ups, yes... That is usually a good thing :)

**Comment by jhprinz** [comment]
So let's do that.

**Comment by dwhswenson** [comment]
Regarding the name used for the shortcut version: I'm concerned that it might be confusing to have both `MinusMover.movers` (the shortcut to the `ConditionalSequentialMover` stuff) and `MinusMover.mover` (the attribute inherited from `WrappedMover`). For a `MinusMover`, the `mover` attribute is structural and essential, so we can't change that. But we can name this shortcut whatever we want.

The advantage to calling the shortcut `movers` is consistency with the `ConditionalSequentialMover` that the `MinusMover` actually uses. The disadvantage is the confusion with `mover`, used by `WrappedMover`.

A different name might avoid that confusion, but then it loses the direct relation to `ConditionalSequentialMover`. If we use a different name, I'd prefer to go far from the "move" theme (so maybe `stages`, rather than `moves`) to really distinguish this shortcut attribute from the structural `mover`.

Anyway, I see a good argument for both sides of this, and I'm not sure which is better. I think I lean toward keeping `movers`, so we always use the same word for the same thing, even if it risks confusion. But if you have other reasons for either approach, let me know.

**Comment by jhprinz** [comment]
I see. Actually. WrappedMover should only have `.submover` and `.submovers` and not only `.mover`.
Would that fix the problem?

**Comment by jhprinz** [comment]
Okay, a little clearer: 

I implemented a tree mixin that is used by PathMoveChange and PathMover. This requires a function that knows about subnodes. And for changes this is called .subchanges or .subchange (if there is only one) and for movers it is .submovers / .submoves.

The use of `.mover` is PathMover specific and can have various meanings. These should be the mover that are substantial to this move and not any submovers that are used for various reasons.

In case of a SequentialMover both are (or can be) the same. Since the submovers are the moves that are specific to the SequentialMove. Yes, and `.submovers` are meant only for traversing the tree. Usually a user will not really use these.

It also means that every mover has a `.submover` property, even Shooters, etc. While only some have a `.mover` property.

So, I would simply change the implementation of `MinusMover` and remove its `.mover` attribute (or hide it) and only have a .submovers / . submover and a .movers property. This will be changed for all `WrappedMover`

**Comment by dwhswenson** [comment]
> So, I would simply change the implementation of MinusMover and remove its .mover attribute (or hide it) and only have a .submovers / . submover and a .movers property. This will be changed for all WrappedMover

Is what you mean here that `WrappedMover` would use its `.submover` in place of its current `.mover`, and `.movers` would be the shortcut to the `ConditionalSequentialMover.movers`?

This doesn't really fix the naming confusion. Make a list of all movers which have submovers (`WrappedMover`s, `SequentialMover`s, etc). Make a list of the name of the submover attributes for them. This should be consistent. And that makes the problem I'm talking about intractable. Because the perfect solution would be to, at the same time, having the following be true for some names of `$wrapped_submove` and `$sequential_submoves` (using `$` to indicate that the actual name of this attribute is what is under discussion):

1.  `MinusMover.$wrapped_submove`, where `$wrapped_submove` is the same name as used in all `WrappedMover`s, etc.
2.  `MinusMover.$sequential_submoves`, where `$sequential_submoves` is the same name as used in the `ConditionalSequentialMover`
3.  `ConditionalSequentialMover.$sequential_submoves` and `WrappedMover.$wrapped_submove` are very similar words, to highlight the similarity
4.  `MinusMover.$wrapped_submove` and `MinusMover.$sequential_submoves` are different enough to avoid confusion.

I don't think these goals are mutually compatible. So there is no perfect solution; we need to find a next-best solution. Honestly, I think the one we have is just fine. We should just highlight the fact that `MinusMover.mover` is a part of the actual structure of the of move decision tree, whereas `MinusMover.movers` is a shortcut to simplify analysis.

To speak to the specific example, if `WrappedMover` uses an attribute called `submover` instead of `mover`, then I think (for consistency) `ConditionalSequentialMover` should use `submovers` in the same way. That's following point 3 above. But we end up with the same problem that if we use `MinusMover.submovers` as a shortcut to the `ConditionaSequentialMover.submovers` (following point 2), it risks confusion between the two names (violating 4). The perfect case of satisfying all 4 of those is, as far as I can tell, impossible.

So we have to break at least one of those rules. With the setup we currently have (well, once I add the attribute) we only violate rule 4. The question becomes which rule is least important, and we violate that one. If 4 is least important, then we can leave it as is.

**Comment by jhprinz** [comment]
Hmmm, let's make this a little clearer. I think the user should only use `.mover` or `.movers`. These are the ones that are Mover specific. There is no need at all to access `.submovers` this is only used (internally) for iterating or if you are interested in the tree structure. So `.submover(s)` is for the tree structure and `.mover(s)` for the user :)

That was my idea. Only "problem" I see is with storage, since it requires to look at the to/from_dict classes to make sure the right objects are saved.

This will also require some reimplementation of the `.movers` property for some movers.

To remove confusion we could even hide the submovers and only access them internally and expose them using iterators, etc...

#### examples

```
# For SequentialMovers both are the same :)
ConditionalSequentialMover.movers = [mover1, mover2, ...]
ConditionalSequentialMover.submovers = [mover1, mover2, ...]

# For MinusMover
MinusMover.movers = [ Truncate, RepEx, Extend ]  # something like this
MinusMover.submovers = [ FilterMover ] # this is for the structure
```

Finally `.canonical` needs the proposed `.is_canonical` attribute to stop canonical from traversing deeper or stopping too early in the tree.

**Comment by dwhswenson** [comment]
The problem is that some user-defined analysis might need knowledge of the move decision tree structure. I can't remember now exactly what I was doing, but something analysis-related is when I noticed the stuff that led to the original post on this issue.

I think that our advanced users (less than contributors, more than novices) are likely to need to work with the move decision tree structure. That's what is behind my concerns about the names of these attributes.

I like the idea of having both available, and under certain circumstances they just happen to be the same. `submovers` for the "truth" of the tree, `movers` for the "simplified conceptual" view. Let's plan that as a PR separate from the `is_canonical` (which already I wrote most of yesterday morning, while avoiding practicing my best man speech).
